src/black/linegen.py

```diff
@@ -329,7 +329,24 @@ def visit_STRING(self, leaf: Leaf) -> Iterator[Line]:
 
             # We could enforce triple quotes at this point.
             quote = quote_char * quote_len
-            leaf.value = prefix + quote + docstring + quote
+
+            if Preview.long_docstring_quotes_on_newline in self.mode:
+                # Put closing quotes on new line if max line length exceeded
+                last_line_length = len(docstring.splitlines()[-1]) if docstring else 0
+
+                # Make the docstring apart from the closing quotes, which happen below
+                docstring = prefix + quote + docstring
+                if (
+                    len(indent) + len(prefix) + 2 * len(quote) + last_line_length
+                    <= self.mode.line_length
+                ):
```

Comment:
I've tried to think this through, but I'm not 100 % still. Wouldn't this be incorrect if we had a multiline docstring whose last line was just acceptable? We would sum the prefix and quote length to it and refuse to add quotes at the end.

```python
def foo():
    """first line---------------------------------------------------------------

    second line--------------------------------------------------------------"""
```

In any case, a test about this would be nice too.

Reply:
You were right. There were two issues with my original code, the last_line_length actually included the indent for docstrings of more than one line. I fixed this issue by using lstrip first. The other issue is that if the docstring has more than one line then we only have one set of quotes so we shouldn't multiply it by two. The logic is a bit more complicated now, but I've added the example above as a test and it works fine now. 

Thanks for the good spot :p

Reply:
Nice 👍 looking at the new code: are you sure that the prefix doesn't affect the calculation similarly? Should we instead only calculate the line length once? Since we only really care about the last one, perhaps something like:

```python
lines = docstring.splitlines()
last_line_len = lines[-1]
if len(lines) == 1:
    last_line_len += quote_len + len(prefix)

if last_line_len + quote_len > max_len:
    # add newline
...
```

I'm not sure how the indentation matters if we strip it only to add it back. Maybe that could simplify it a bit. But I might be disregarding some details.

Reply:
I was pretty tired when working on this last night, that's my excuse for missing the prefix :p 
I think it's all sorted now. I've added some tests with prefixes now just to be sure. I've also refactored it how you suggest and added in some (possibly excessive) comments to explain exactly why we're doing all this.

Thanks for your help :)

Reply:
No worries! Thanks for being so quick with the edits 👍